### ISSUE TITLE
'operator=' has a non-throwing exception specification but can still throw

### ISSUE THREAD

**Issue by mobileben** [Tencent/rapidjson#1280]
I'm using rapidjson for a game running iOS. I'm getting complier warnings related to rapidjson

> 'operator=' has a non-throwing exception specification but can still throw [-Wexceptions]
> 'GenericValue' has a non-throwing exception specification but can still throw [-Wexceptions]

The issue at hand is in code such as this:

    GenericValue& operator=(GenericValue& rhs) RAPIDJSON_NOEXCEPT {
        RAPIDJSON_ASSERT(this != &rhs);
        this->~GenericValue();
        RawAssign(rhs);
        return *this;
    }

The problem is I can't modify ```RAPIDJSON_ASSERT``` to "do nothing" if ```RAPIDJSON_NOEXCEPT``` becomes noexcept since ```RAPIDJSON_ASSERT``` is used in other contexts independent of ```RAPIDJSON_NOEXCEPT```.

It seems perhaps there needs to be a variant (terrible name warning) like ```RAPIDJSON_ASSERT_NOEXCEPT``` that can turn itself off when ```clang``` and ```__has_feature(cxx_noexcept)``` is true

I just pulled the latest and the issue still exists.

Note that this issue also surfaced here:

https://github.com/USCiLab/cereal/issues/439

**Comment by miloyip** [comment]
I think it could be safely changed to

~~~cpp
    if (this != &rhs) {
        this->~GenericValue();
        RawAssign(rhs);
    }
    return *this;
~~~

**Comment by mobileben** [comment]
I'll update and create a pull request. There are still a couple of other instances of this:

    explicit GenericValue(Type type) RAPIDJSON_NOEXCEPT : data_() {
        static const uint16_t defaultFlags[7] = {
            kNullFlag, kFalseFlag, kTrueFlag, kObjectFlag, kArrayFlag, kShortStringFlag,
            kNumberAnyFlag
        };
        RAPIDJSON_ASSERT(type >= kNullType && type <= kNumberType);
        data_.f.flags = defaultFlags[type];

        // Use ShortString to store empty string.
        if (type == kStringType)
            data_.ss.SetLength(0);
    }

and

```
#ifndef RAPIDJSON_DOXYGEN_RUNNING // hide SFINAE from Doxygen
    template <typename T>
    explicit GenericValue(T b, RAPIDJSON_ENABLEIF((internal::IsSame<bool, T>))) RAPIDJSON_NOEXCEPT  // See #472
#else
    explicit GenericValue(bool b) RAPIDJSON_NOEXCEPT
#endif
        : data_() {
            // safe-guard against failing SFINAE
            RAPIDJSON_STATIC_ASSERT((internal::IsSame<bool,T>::Value));
            data_.f.flags = b ? kTrueFlag : kFalseFlag;
    }
```

**Comment by miloyip** [comment]
The 2nd one should be fine since static assert is a compile-time thing.
The 1st one is quite troublesome though... I have no idea how to solve it.

**Comment by mobileben** [comment]
Some options ... not saying they are good. But just offering out ideas good or bad to see if we can figure out how to reasonably handle this.

I get the intent for the testing around the type value. I suppose he question is when and how the user learns of issues. On one hand you can only protect the user so much. 

- Just allow this version of the constructor to throw
- Create a macro like RAPIDJSON_ASSERT_NOEXCEPT or RAPIDJSON_NOEXCEPT_ASSERT which will effectively disable the throw for clang/cxx_noexcept. The macro wouldn't have to be defined by the user. Rather the lib would either make it use RAPIDJSON_ASSERT or be "empty"
-Introduce a DEBUG flag to manage this
-Treat failure as if it was a kNullType (which will cause problems later on for the user)
-Introduce a flag to hold this state (this seems like a total waste to do since it isn't clear how often people assign an illegal type value).
-Just allow the warnings to exist
-Moved to scoped enums for C++11

As I said some bad suggestions.

**Comment by mobileben** [comment]
I plan on doing something like this:

+///////////////////////////////////////////////////////////////////////////////
+// RAPIDJSON_NOEXCEPT_ASSERT
+
+#if RAPIDJSON_HAS_CXX11_NOEXCEPT
+#define RAPIDJSON_NOEXCEPT_ASSERT(x)
+#else
+#define RAPIDJSON_NOEXCEPT_ASSERT(x) RAPIDJSON_ASSERT(x)
+#endif // RAPIDJSON_HAS_CXX11_NOEXCEPT
+

However, I totally understand your point of assert v. throw. One thing I also thought of ... perhaps have a user defined of RAPIDJSON_ASSERT_THROWS or RAPIDJSON_ASSERT_USES_THROW

This way we could use that to control whether or not RAPIDJSON_NOEXCEPT_ASSERT will actually become nothing. This way it would still protect anyone that uses assert(x) versus throw?

Basically something like this:

#ifdef RAPIDJSON_ASSERT_THROWS
#if RAPIDJSON_HAS_CXX11_NOEXCEPT
#define RAPIDJSON_NOEXCEPT_ASSERT(x)
#else
#define RAPIDJSON_NOEXCEPT_ASSERT(x) RAPIDJSON_ASSERT(x)
#endif // RAPIDJSON_HAS_CXX11_NOEXCEPT
#else
#define RAPIDJSON_NOEXCEPT_ASSERT(x) RAPIDJSON_ASSERT(x)
#endif // RAPIDJSON_ASSERT_THROWS
